iothub/device/src/InternalClient.cs

```diff
@@ -1468,11 +1402,7 @@ private async Task OnDeviceMessageReceivedAsync(Message message)
 
             try
             {
-                // Wait to acquire the _deviceReceiveMessageSemaphore. This ensures that we get a reference to the latest callback set on the device client.
-                await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
```

Comment:
This lock shouldn't be removed. We want to ensure that the callback is invoked on the latest supplied handler.

Reply:
Even with this lock, the handler could get updated immediately after this value is read. I can remove this for now since it isn't critical for fixing the github issue, but I don't see the value of leaving it.

Reply:
I agree, it could, but since we invoke it immediately after, it does reduce the scope of us invoking the handler on an updated callback. Also, you could release the semaphore after invoking the callback, that way a user won't be able to update the callback in the interim.

Reply:
I'll move the semaphore here to fully wrap the scope of the read and invoke of the callback then, sure